sdk/cosmos/azure-cosmos/azure/cosmos/_session_token_helpers.py

```diff
@@ -0,0 +1,224 @@
+# [("AA", "DD"), "1:1#57,2:1#58"]
+# 3. [(("AA", "BB"), "1:1#57"), (("BB", "DD"), "2:1#52"), (("AA", "DD"), "0:1#55")] ->
+# [("AA", "DD"), "1:1#57,2:1#52,0:1#55"]
+# goal here is to detect any obvious merges or splits that happened
+# compound session tokens are not considered will just pass them along
+def merge_ranges_with_subsets(overlapping_ranges: List[Tuple[Range, str]]) -> List[Tuple[Range, str]]:
+    processed_ranges = []
+    while len(overlapping_ranges) != 0: # pylint: disable=too-many-nested-blocks
+        feed_range_cmp, session_token_cmp = overlapping_ranges[0]
+        # compound session tokens are not considered for merging
+        if is_compound_session_token(session_token_cmp):
+            processed_ranges.append(overlapping_ranges[0])
+            overlapping_ranges.remove(overlapping_ranges[0])
+            continue
+        _, vector_session_token_cmp = parse_session_token(session_token_cmp)
+        subsets = []
+        # finding the subset feed ranges of the current feed range
+        for j in range(1, len(overlapping_ranges)):
+            feed_range = overlapping_ranges[j][0]
+            if not is_compound_session_token(overlapping_ranges[j][1]) and \
+                    feed_range.is_subset(feed_range_cmp):
+                subsets.append(overlapping_ranges[j] + (j,))
+
+        # go through subsets to see if can create current feed range from the subsets
+        not_found = True
+        j = 0
+        while not_found and j < len(subsets):
+            merged_range = subsets[j][0]
+            session_tokens = [subsets[j][1]]
+            merged_indices = [subsets[j][2]]
+            if len(subsets) == 1:
+                _, vector_session_token = parse_session_token(session_tokens[0])
+                if vector_session_token_cmp.global_lsn > vector_session_token.global_lsn:
+                    overlapping_ranges.remove(overlapping_ranges[merged_indices[0]])
+            else:
+                for k, subset in enumerate(subsets):
+                    if j == k:
+                        continue
+                    if merged_range.can_merge(subset[0]):
+                        merged_range = merged_range.merge(subset[0])
+                        session_tokens.append(subset[1])
+                        merged_indices.append(subset[2])
+                    if feed_range_cmp == merged_range:
+                        # if feed range can be created from the subsets
+                        # take the subsets if their global lsn is larger
+                        # else take the current feed range
+                        children_more_updated = True
+                        parent_more_updated = True
+                        for session_token in session_tokens:
+                            _, vector_session_token = parse_session_token(session_token)
+                            if vector_session_token_cmp.global_lsn > vector_session_token.global_lsn:
```

Comment:
I think here we assume if GLSN of _enclosing feed range_ is greater then corresponding LLSNs are also greater in the _enclosing feed range_.